### ISSUE TITLE
RFC: per-submission credit

### ISSUE THREAD

**Issue by mwest1066** [PrairieLearn/PrairieLearn#10607]
**Your feedback is wanted!** _This is a tentative plan for changing the way "credit" works, both bonus credit (above 100%) and reduced credit (below 100%). I'm writing this up for the purpose of collecting feedback so please comment on this issue! In particular, I'd love to hear whether you use different "credit" at the moment and whether you think this change would be helpful for you or not._

## Goals

At the moment, "credit" operates on a whole-assessment-instance basis. This is confusing for bonus credit (above 100%), because it means that a student needs to exceed 100% during the extra-credit time to get any benefit from it. It also means that reduced credit can fail to give an incentive to continue working. For example, if a homework has 50% late credit, a student who already achieved 50% on the assessment has no incentive to continue working during the late period.

To fix both of these issues, we should award bonus credit on a more granular level. The best case would be do this per-submission.

The main difficulty is what to do with things like "maximum points", which can be set per-question, per-zone, or per-assessment. A secondary issue is how to display all of this to the user (both student and instructor). We make a proposal below for both of these issues.

## Proposed solution

The basic idea is to track two different types of points:
* `base_points`: these are the points that would be awarded if credit wasn't taken into account.
* `credit_points`: these are the `base_points` scaled by `credit_perc`.

The key insight is that `max_points` apply to `base_points`. So we add up the `base_points` until we get to `max_points` and then we are capped. At the same time we add up the associated `credit_points` to get the total, which is the number of points actually earned for the assessment.

## Example calculations

Suppose we have earned:
* 4 points at 120% (this means 4 `base_points` and 4 * 1.2 = 4.8 `credit_points`)
* 3 points at 100% (this means 3 `base_points` and 3 `credit_points`)
* 6 points at 80% (this means 6 `base_points` and 6 * 0.8 = 4.8 `credit_points`)

If `max_points = 4`:
* We use the first set of 4 `base_points`, which have a corresponding 4 * 1.2 = 4.8 `credit_points`
* total awarded points = 4.8 `credit_points`

If `max_points = 6`:
* 4 `base_points` at 120% are used, giving a total of 4 `base_points` and 4.8 `credit_points`
* 2 `base_points` at 100% are used, giving a total of 6 `base_points` and 6.8 `credit_points`
* We are now capped out at our 6 `max_points` (based on the total `base_points`)
* total awarded points = 6.8 `credit_points`

If `max_points = 8`:
* 4 `base_points` at 120% are used, giving a total of 4 `base_points` and 4.8 `credit_points`
* 3 `base_points` at 100% are used, giving a total of 7 `base_points` and 7.8 `credit_points`
* 1 `base_points` at 80% are used, giving a total of 8 `base_points` and 8.6 `credit_points`
* We are now capped out at our 8 `max_points` (based on the total `base_points`)
* total awarded points = 8.6 `credit_points`

If `max_points = 12`:
* 4 `base_points` at 120% are used, giving a total of 4 base points and 4.8 `credit_points`
* 3 `base_points` at 100% are used, giving a total of 7 base points and 7.8 `credit_points`
* 5 `base_points` at 80% are used, giving a total of 12 base points and 11.8 `credit_points`
* We are now capped out at our 12 `max_points` (based on the total `base_points`)
* Note that we actually have less than 12 awarded points
* total awarded points = 11.8 `credit_points`

If `max_points = 15`:
* 4 `base_points` at 120% are used, giving a total of 4 base points and 4.8 `credit_points`
* 3 `base_points` at 100% are used, giving a total of 7 base points and 7.8 `credit_points`
* 6 `base_points` at 80% are used, giving a total of 13 base points and 12.6 `credit_points`
* We still have 2 `base_points` available from our total `max_points` so we aren't capped
* total awarded points = 12.6 `credit_points`

## Submissions

Each submission stores the credit and two types of points:
* `credit_perc`
* `base_points`
* `credit_points = base_points * credit_perc / 100`

When displaying a submission with `base_points = 4` and `credit_perc = 120` to students and instructors, we write it as:
* When displaying a summary list, such as the list of points of previous submissions, we display only `credit_points`
* On the submission itself, we display `4.8 credit points (4 points at 120% credit)`

## Instance questions

Each instance question stores:
* `credit_points`: the total credit points awarded for this question
* `base_points_list`: a list of the form `[[base_points_1, credit_perc_1], [base_points_2, credit_perc_2], ...]`

To compute the instance question points from the submissions we take the `[base_points, credit_points]` from each submission and combine them into a single `base_points_list` array. Then use the points-cap algorithm below to produce the `credit_points` and `base_points_list` for the instance question.

When displaying the points for an instance question, we display the `credit_points` as a plain number with a `[?]` icon next to it that triggers a popover, which then displays something like:
* `4.8 credit points (4 points at 120% credit)`
* `3 credit points (3 points at 100% credit)`
* `1.6 credit points (2 points at 80% credit)`
* `Total credit points = 9.4`

## Zones

Each zone has an effective `base_points_list`. To combine the points together from the questions within the zone:
1. If the zone has `bestQuestions` set, then select this many questions with the highest `credit_points` values, otherwise select all questions.
2. Concatenate the `base_point_list` from the selected questions together, then use the points-cap algorithm below (with the zone `max_points`) to produce the combined `base_points_list` for the zone.

## Assessment instances

An assessment instance also stores:
* `credit_points`: the actual points awarded for this assessment instance
* `base_points_list`
* `score_perc`

To compute the points, concatenate the `base_points_list` for the zones together and call 

## Points-cap algorithm

This algorithm reduces an `input_base_points_list` to an `output_base_points_list` by capping the maximum base points at a given `max_points` value. It also computs the total `credit_points` associated with the `output_base_points_list`.

1. Inputs:
    1. `input_base_points_list` of the form `[[base_points_1, credit_perc_1], [base_points_2, credit_perc_2], ...]`
    2. `max_points` value for the combined points
2. Outputs:
    1. `credit_points`: the total credit points awarded for the combined objects
3. Initialize:
    1. `max_points_remaining = max_points`
    2. `credit_points = 0`
    3. `output_base_points_list = []`
4. For each `[base_points, credit_perc]` in `input_base_points_list` in descending order of `credit_perc`:
    1. `useable_base_points = min(max_points_remaining, base_points)`
    2. If `output_base_points_list.length > 0` and `output_base_points_list[-1][1] == credit_perc`:
        1. `output_base_points_list[-1][0] += useable_base_points`
    3. Else:
        1. `output_base_points_list.push([useable_base_points, credit_perc])`
5. For each `[base_points, credit_perc]` in `output_base_points_list`:
    1. `credit_points += base_points * credit_perc / 100`

## Manual grading

We actually have `max_auto_points` for instance questions, rather than `max_points`. We are treating manual points as assigned directly to the question, not to an individual submission. Does this mean that we don't want to consider credit for manual grading? It does seem to me that it would be nice if we could handle the case where a student submits a manually-graded answer late and is automatically awarded 80% for it. Maybe we should have an explicit `credit_perc` associated with manual grading scores, which default to the credit of the most recent submission at the time of manual grading, but which can also be set by the grader?

## Handling edits by instructors

What do we do when an instructor manually edits an instance question score? That is, not manual grading, but just directly editing the total points awarded.

Option 1:
* For each manual edit, store the manual edit object with the date and the number of points assigned. Then when computing `base_points_list` only include the latest manual edit, treating it as `[edited_value, 100]` and any submissions after the date of the last manual edit. We could even store a marker in `base_points_list` to indicate that this entry was a manual edit so that we can display this information when clicking the icon for detailed points info.

Option 2:
* The algorithms above were written assuming that we recalculate all scores (instance question, zone, assessment instance) from scratch each time any submission is made. We could instead do this incrementally, which is more like how we do it at the moment with instance questions. This would mean that we are effectively not sorting `input_base_points_list` by descending `credit_perc`, but instead just taking everything in date order. If a student has made some late submissions at 80% credit, but then is given (or buys with tokens) an extension at 100% credit, they wouldn't be able to achieve max points because they'd already be locked in with their 80% submissions.

## Handling `max_bonus_points`

An assessment can have `max_bonus_points` set, as well as `max_points`. In this case we just use the sum of the two as the real maximum, but then divide by `max_points` to get the `score_perc`. This shouldn't require any changes to the above algorithms.

## Regrading

If we go with the "recompute points from scratch" strategy outlined above, regrading would be quite simple because we'd just reprocess all questions and then the overall assessment instance.

## Questions

1. What should we do about manual grading and credit (see comments above)?
2. Which option (or other alternative) should we use for handling score edits by instructors?
3. We really should be storing a per-student zone structure so we can handle the case when an exam zone configuration is changed after the student starts the exam. We should then store `credit_points`/`base_points_list` in the zone.

**Comment by echuber2** [comment]
Just for sake of discussion, there is currently a hacky way to cause a scaling factor to always be applied to the score: https://github.com/PrairieLearn/PrairieLearn/issues/11291

**Comment by reteps** [comment]
Replaces #818

**Comment by reteps** [comment]
We discussed that we probably don't want backwards compatible behavior here, we would just redefine credit to have this better, more sensible behavior.

**Comment by mwest1066** [comment]
> We discussed that we probably don't want backwards compatible behavior here, we would just redefine credit to have this better, more sensible behavior.

That's correct. We won't attempt to maintain backwards compatibility with the existing behavior. This new system will use the existing `credit` settings, but will implement them in this new way. The new system will be essentially what most people always thought `credit` meant in the first place. I don't think anyone will be relying on the exact current behavior, but in case someone is we will announce this change well in advance, probably implement it via an opt-in per-course feature flag initially, and only later will we drop the old behavior.
